(#69) (#85) 운동기록 수정페이지 개발, 인풋필드 유효성 검증 및 기술부채의 오류 해결

ldg-96 opened the pull request:
## #️⃣연관된 이슈
>#69, #85

## 📝작업 내용
- (#69)
  - 운동기록 수정 페이지 개발
  - 운동기록 생성,수정에서 사용할수 있도록 exerciseForm컴포넌트 수정
  
- (#85) (taskList의 내용 개발)
  - exerciseForm 인풋필드 유효성 검증 추가
  - 타입정의시 type으로 통일
  - 인풋필드에서 엔터키 입력시 제출되는 이벤트 막기
  - 운동기록 생성시 시작시간, 종료시간에 9시간 offset 넣기

- 기타
  - 운동기록 리스트(list), 생성(create), 상세(detail), 수정(edit) 페이지의 page.tsx파일을 Presentation & Container 패턴이 되지 않도록 변경

> - useApiErrorHandler 훅은 만들어 컴포넌트에서 api사용시 에러핸들링 할 수 있도록 하였습니다. 현재는 전체적으로 공통 오류인 401만 넣어두고 나머지는 사용하는 컴포넌트에서 추가적으로 받도록 하였고 추후 공통적으로 처리할 오류가 늘어나면 더 추가할 예정입니다.
>- 운동기록 페이지들의 url 쿼리파라미터 유효성 검증을 현재는 각 페이지에서 아래와 같은 방식으로 하고 있습니다. 혹시 추천해주시는 방법이 있을까요?
```
  useEffect(() => {
    if (!exerciseId) {
      router.replace(SITE_MAP.EXERCISE)
    }
  }, [exerciseId])
```

ldg-96 committed:
2025-06-12

ldg-96 committed:
2025-06-12

ldg-96 committed:
Merge branch 'refactor/api' of github.com:projects200/front-server into feature/exercise-edit

# Conflicts:
#	src/app/exercise/_components/exerciseForm.tsx

ldg-96 committed:
2025-06-12

ldg-96 committed:
2025-06-13

ldg-96 committed:
Merge branch 'refactor/api' of github.com:projects200/front-server into feature/exercise-edit

# Conflicts:
#	src/app/exercise/_components/exerciseForm.tsx

ldg-96 committed:
cherry-pick

ldg-96 committed:
Merge branch 'develop' of github.com:projects200/front-server into feature/exercise-edit

ldg-96 committed:
2025-06-13

ldg-96 committed:
2025-06-13

ldg-96 committed:
(#69) (#85) 운동기록 수정페이지 개발, 인풋필드 유효성 검증 및 taskList의 오류 해결

SimYunSup commented on src/hooks/useExerciseApi.ts:
```diff
@@ -37,8 +38,8 @@ export const usePostExercisePictures = () =>
     ExercisePicturesUpload & { exerciseId: number }
-  >(['exercise/pictures'], (token, { images, exerciseId }) =>
-    createExercisePictures(token, { images }, exerciseId),
+  >(['exercise/create/pictures'], (token, { newImages, exerciseId }) =>
```
**question:** key를 위와 같이 구성한 이유라도 있을까요?

ldg-96 commented on src/app/exercise/edit/page.tsx:
```diff
@@ -1,33 +1,110 @@
+
+  const loading = isLoading || isPatching || isDeletingPics || isUploadingPics
+
+  const defaultValues = useMemo(() => {
+    if (!data) return null
+    const { title, category, location, startedAt, endedAt, content } = data
+    return { title, category, location, startedAt, endedAt, content } as const
+  }, [data])
```
```
  const defaultValues =
    data === undefined
      ? null
      : ({
          title: data.title,
          category: data.category,
          location: data.location,
          startedAt: data.startedAt,
          endedAt: data.endedAt,
          content: data.content,
        } as const)
```
tanstack/react-form은 부모가 다시 리렌더되도 폼을 reset() 하지 않았습니다. 하여 useMemo를 제거하고 위와같이 코드를 변경하였습니다.

SimYunSup replied:
`defaultValue`를 새로 선언해서 사용하는 것과 `data`를 그대로 사용하는 것과 어떤 차이가 있죠?

ldg-96 replied:
불필요한 변수선언인것 같습니다.
```
 <ExerciseForm
        defaultValues={{
          title: data?.title,
          category: data?.category,
          startedAt: data?.startedAt,
          endedAt: data?.endedAt,
          location: data?.location,
          content: data?.content,
        }}
        defaultPictures={data?.images}
        onSubmit={handleSubmit}
        onError={(message) => showToast(message, 'info')}
      />
```
위와같이 직접 ExerciseForm컴포넌트에 data를 삽입하도록 변경하였습니다.

ldg-96 commented on src/app/exercise/list/page.tsx:
```diff
@@ -6,39 +6,38 @@ import { useRouter } from 'next/navigation'
   useEffect(() => {
-    if (!error) return
-    const navigated = handleApiError(error)
-    if (!navigated) router.back()
+    if (error) {
+      handleError(error, {
+        actions: { 400: 'back' },
+      })
+    }
   }, [error])
```
위 내용과 동일합니다.

ldg-96 commented on src/app/exercise/detail/page.tsx:
```diff
@@ -5,42 +5,62 @@ import { useRouter } from 'next/navigation'
+        actions: { 403: 'back', 404: 'back' },
+      })
+    }
+  }, [error])
```
위 내용과 동일합니다.

ldg-96 committed:
(#69) (#85) 운동기록 수정페이지 개발, 인풋필드 유효성 검증 및 taskList의 오류 해결
수정내용 : useMemo 삭제, 각 페이지에서 하던 에러핸들링을 useExerciseApi 훅에서 하도록 변경

SimYunSup commented on src/app/exercise/_components/exerciseForm.tsx:
```diff
@@ -12,44 +13,63 @@ import TextareaField from './textareaField'
+          '운동 상세 내용은 최대 65,535바이트까지 입력 가능합니다.',
+        )
+        .optional(),
+      newImages: z.array(z.instanceof(File)).max(remain, '이미지는 최대 5장까지 업로드할 수 있습니다.'),
```
**question(blocking):** `z.union([z.instanceof(File), z.string()])으로 두면 안되는건가요?

ldg-96 replied:
```
 images: z
        .array(
          z.union([
            z
              .instanceof(File)
              .refine(
                (file) => file.size <= 10 * 1024 * 1024 && ['image/jpeg', 'image/png', 'image/jpg'].includes(file.type),
                'jpg, jpeg, png 파일만 가능하며, 용량은 10MB 이하만 업로드할 수 있습니다.',
              ),
            z.string().url('유효하지 않은 이미지 URL입니다.'),
          ]),
        )
        .max(5, '이미지는 최대 5개까지 업로드할 수 있습니다.'), 
```
조언해주신대로 코드를 변경해 보았습니다.
`useState`로관리하던 `existingPictures, deletedIds`,  `useMemo`로 관리하던 `remain`값을 제거하였고 대신에
newImages가 아닌 images의 타입을 `(File | string)[]` 로 두어 파일이미지와 기존이미지(url)을 관리할 수 있도록 하였습니다.


```
    onSubmit: ({ value }) => {
      const deletedIds = defaultPictures
        .filter((picture) => !value.images.includes(picture.pictureUrl))
        .map((picture) => picture.pictureId)

      const newFiles = value.images.filter((v): v is File => v instanceof File)

      const payload: ExerciseRecordReq = {
        ...value,
        deletedIds,
        images: newFiles,
      }

      onSubmit(payload)
    },

```
제출시 삭제할 사진의id는 기존에 있던 배열의 url에서 사라진 url의 id를 찾아 넣어주었고,
새로운 사진은 배열의 값들중 타입이 file인 것을 찾아 넣어주어 제출하였습니다.

ldg-96 commented on src/app/exercise/edit/page.tsx:
```diff
@@ -1,33 +1,110 @@
+        endedAt: value.endedAt,
+      })
+    } catch (error) {
+      handleError(error)
```
handleError를 useApiGet, useApiMutation에 추가하고 

```
export const useReadExerciseList = (date: string) =>
  useApiGet<ExerciseList[]>(
    ['exercise/list', date],
    (token) => readExerciseList(token, date).then(adaptExerciseList),
    {
      revalidateOnFocus: false,
      shouldRetryOnError: false,
    },
    {
      messages: {
        400: '올바르지 않은 날짜입니다.',
      },
      actions: { 400: 'back' },
    },
  )

```
위 코드와 같이 에러상황에 관해서는 useExerciseApi훅에서 일괄적으로 관리하고자 이렇게 구현해 보았습니다. 만약 추후 useExerciseApi훅을 사용하는 컴포넌트에서 onError를 사용해야할 일이 생긴다면

```
return useSWR<Data, ApiError>(key, fetcher, {
  ...options,
  onError: (err, key, config) => {
    options?.onError?.(err, key, config)
    handleError(err, policy)
  },
})
```
이런식으로 useApiGet, useApiMutation훅을 상위컴포넌트에서 넘긴 에러와 handleError를 병합하고

```
export const useReadExerciseList = (
  date: string,
  options?: SWRConfiguration<ExerciseList[], ApiError>,
) =>
  useApiGet<ExerciseList[]>(
    ['exercise/list', date],
    token => readExerciseList(token, date).then(adaptExerciseList),
    {
      revalidateOnFocus: false,
      shouldRetryOnError: false,
      // 상위 컴포넌트에서 받은 옵션 추가
      ...options, 
    },
    {
      messages: {
        400: '올바르지 않은 날짜입니다.',
      },
      actions: { 400: 'back' },
    },
  )

```
위와같은 방식으로 확장할 수 있을것 같습니다. 이러한 방식도 괜찮을까요?

SimYunSup replied:
네 괜찮습니다. 하지만 추상화는 비용이 추가로 들기 때문에, 여기서 message와 action에서 다른 기능을 추가하지는 않았으면 합니다.

ldg-96 replied:
넵 확인하였습니다!

ldg-96 committed:
(#69) (#85) 운동기록 수정페이지 개발, 인풋필드 유효성 검증 및 taskList의 오류 해결
수정내용 : defaultValue를 선언하지않고 ExerciseForm컴포넌트에 데이터 바로 삽입, z.union() 을 사용하여 file, string타입의 이미지를 관리하도록 변경
